File: service/Api/src/Shared/Application/Models/Results/Result.cs
using System.Text.Json.Serialization;

namespace Shared.Application.Models.Results;

public readonly partial record struct Result : IResultRecord
{
    #region Properties
    public bool IsSuccess { get; }

    public int StatusCode { get; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Message { get; }

    public List<Error> Errors { get; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public IReadOnlyDictionary<string, object?>? Metadata { get; }
    #endregion

    #region Compute
    [JsonIgnore]
    public bool IsFailure => !IsSuccess;
    #endregion

    #region Constructors
    [JsonConstructor]
    internal Result(
        bool isSuccess = ResultConstant.DefaultValues.IsSuccess,
        int statusCode = ResultConstant.DefaultValues.StatusCode,
        string? message = null,
        List<Error>? errors = null,
        IReadOnlyDictionary<string, object?>? metadata = null)
    {
        IsSuccess = isSuccess;
        StatusCode = statusCode;
        Message = message;
        Errors = errors ?? [];
        Metadata = metadata;
    }
    #endregion
}

File: app/Admin/src/shared/models/result.ts
export interface Error {
  code: string
  message: string
  type: number
  metadata: Record<string, unknown> | null
}

export interface Result<T> {
  isSuccess: boolean
  statusCode: number
  errors: Error[]
  message: string | null
  metadata: Record<string, unknown> | null
  value: T
}

export interface PagedResult<T> {
  isSuccess: boolean
  statusCode: number
  errors: Error[]
  message: string | null
  metadata: Record<string, unknown> | null
  items: T[]
  page: number
  pageSize: number
  totalCount: number
}
